enhancements/cluster-scope-secret-volumes/csi-driver-host-injections.md

```diff
@@ -0,0 +1,511 @@
+ - the example in the repo's README shows how you can write data to the volume by exec'ing into the application Pod,
+ and then similarly see that same data if you exec into the CSI driver Hostpath container.
+ - But instead of this, in our case, we would build a co-located (in the same process as the CSI driver Hostpath container)
+ an operator/controller that watches the secret(s) we want to mount
+ - the associated Listers for the Operator/Controller would be made available to the CSI plugin code
+ - And then in methods like `createHostpathVolume`, in addition to creating the directory, access the Secret(s) from the
+ Lister and serialize the secret(s) bytes into files off of the directory the CSI plugin maintains
+ - However, exactly when we create directories for the mount and update the secrets files does not have to be at the same time.
+ - We will minimize the writes of the Secret to disk, so that is does not occur every time we provision a Pod
+ - Per [review of this proposal](https://github.com/openshift/enhancements/pull/250/files#r396278957), the CSI driver
+ is not limited by when CSI calls between K8S and itself, occur.  So we don't have to populate the disk as part of 
+ any given call.  As long as the data is there by the time the final provisioning occurs, we are good.
+
+The [secrets store csi plugin](https://github.com/kubernetes-sigs/secrets-store-csi-driver) as very similar to hostpath
+in that 
+- the CSI spec implementation points were easy to find
+- there were golang file system level calls during volume provisioning, i.e. `NodePublishVolume`, like `ioutil.ReadDir`
+- and use of the k8s util mount API's
+- it already employs `DaemonSets`
+- the `SecretProviderClass` examples did not work out of the box on OpenShift, but we are not going down that path 
+anyway.
+
+
+### Risks and Mitigations
+
+While this proposal settles on installing via CVO, that decision has ramifications whose discussion are noted above.
+It is at least conceivable the need to revisit that decision may arise when implementation starts.
+
+Getting the flow and various stages from the CSI specification understood and correctly utilized so that we minimize
+the amount of times the Secret is read and stored on the local filesystem of each host/node is important.  We want
+to minimize local file system writes to when the Secret actually changes, and not when we provision the volume to 
+a new Pod. 
+
+
+## Design Details
+
+### Test Plan
+
+Verification of this proposal should be containable around e2e's that automate various forms of [the basic Hostpath
+driver confirmation](https://github.com/kubernetes-csi/csi-driver-host-path/blob/master/docs/deploy-1.17-and-later.md#confirm-hostpath-driver-works), where that
+confirmation is augmented to consider:
+
+- the multiple replicas of the `DaemonSet`
+- create/update/delete of the K8S Secret(s) designated for host/node injection are properly reflected on the volume
```

Comment:
Please explicitly add a test for RBAC revocation (i.e. user looses access to ProjectedResource while running a pod that uses it)

Reply:
part of next push